File: src/main/java/emu/grasscutter/game/entity/GameEntity.java
package emu.grasscutter.game.entity;

import emu.grasscutter.data.GameData;
import emu.grasscutter.game.ability.*;
import emu.grasscutter.game.player.Player;
import emu.grasscutter.game.props.*;
import emu.grasscutter.game.world.*;
import emu.grasscutter.net.proto.FightPropPairOuterClass.FightPropPair;
import emu.grasscutter.net.proto.GadgetInteractReqOuterClass.GadgetInteractReq;
import emu.grasscutter.net.proto.MotionInfoOuterClass.MotionInfo;
import emu.grasscutter.net.proto.MotionStateOuterClass.MotionState;
import emu.grasscutter.net.proto.SceneEntityInfoOuterClass.SceneEntityInfo;
import emu.grasscutter.net.proto.VectorOuterClass.Vector;
import emu.grasscutter.scripts.data.controller.EntityController;
import emu.grasscutter.server.event.entity.*;
import emu.grasscutter.server.packet.send.PacketEntityFightPropUpdateNotify;
import it.unimi.dsi.fastutil.ints.*;
import java.util.*;
import lombok.*;

public abstract class GameEntity {
    @Getter private final Scene scene;
    @Getter protected int id;
    @Getter @Setter private SpawnDataEntry spawnEntry;

    @Getter @Setter private int blockId;
    @Getter @Setter private int configId;
    @Getter @Setter private int groupId;

    @Getter @Setter private MotionState motionState;
    @Getter @Setter private int lastMoveSceneTimeMs;
    @Getter @Setter private int lastMoveReliableSeq;

    @Getter @Setter private boolean lockHP;

    @Setter(AccessLevel.PROTECTED)
    @Getter
    private boolean isDead = false;

    // Lua controller for specific actions
    @Getter @Setter private EntityController entityController;
    @Getter private ElementType lastAttackType = ElementType.None;

    @Getter private List<Ability> instancedAbilities = new ArrayList<>();

    @Getter
    private Int2ObjectMap<AbilityModifierController> instancedModifiers =
            new Int2ObjectOpenHashMap<>();

    @Getter private Map<String, Float> globalAbilityValues = new HashMap<>();

    public GameEntity(Scene scene) {
        this.scene = scene;
        this.motionState = MotionState.MOTION_STATE_NONE;
    }

    public abstract void initAbilities();

    public EntityType getEntityType() {
        return EntityIdType.toEntityType(this.getId() >> 24);
    }

    public abstract int getEntityTypeId();

    public World getWorld() {
        return this.getScene().getWorld();
    }

    public boolean isAlive() {
        return !this.isDead;
    }

    public LifeState getLifeState() {
        return this.isAlive() ? LifeState.LIFE_ALIVE : LifeState.LIFE_DEAD;
    }

    public abstract Int2FloatMap getFightProperties();

    public abstract Position getPosition();

    public abstract Position getRotation();

    public void setFightProperty(FightProperty prop, float value) {
        this.getFightProperties().put(prop.getId(), value);
    }

    public void setFightProperty(int id, float value) {
        this.getFightProperties().put(id, value);
    }

    public void addFightProperty(FightProperty prop, float value) {
        this.getFightProperties().put(prop.getId(), this.getFightProperty(prop) + value);
    }

    public float getFightProperty(FightProperty prop) {
        return this.getFightProperties().getOrDefault(prop.getId(), 0f);
    }

    public boolean hasFightProperty(FightProperty prop) {
        return this.getFightProperties().containsKey(prop.getId());
    }

    public void addAllFightPropsToEntityInfo(SceneEntityInfo.Builder entityInfo) {
        this.getFightProperties()
                .forEach(
                        (key, value) -> {
                            if (key == 0) return;
                            entityInfo.addFightPropList(
                                    FightPropPair.newBuilder().setPropType(key).setPropValue(value).build());
                        });
    }

    protected MotionInfo getMotionInfo() {
        return MotionInfo.newBuilder()
                .setPos(this.getPosition().toProto())
                .setRot(this.getRotation().toProto())
                .setSpeed(Vector.newBuilder())
                .setState(this.getMotionState())
                .build();
    }

    public float heal(float amount) {
        return heal(amount, false);
    }

    public float heal(float amount, boolean mute) {
        if (this.getFightProperties() == null) {
            return 0f;
        }

        float curHp = this.getFightProperty(FightProperty.FIGHT_PROP_CUR_HP);
        float maxHp = this.getFightProperty(FightProperty.FIGHT_PROP_MAX_HP);

        if (curHp >= maxHp) {
            return 0f;
        }

        float healed = Math.min(maxHp - curHp, amount);
        this.addFightProperty(FightProperty.FIGHT_PROP_CUR_HP, healed);

        this.getScene()
                .broadcastPacket(
                        new PacketEntityFightPropUpdateNotify(this, FightProperty.FIGHT_PROP_CUR_HP));

        return healed;
    }

    public void damage(float amount) {
        this.damage(amount, 0, ElementType.None);
    }

    public void damage(float amount, ElementType attackType) {
        this.damage(amount, 0, attackType);
    }

    public void damage(float amount, int killerId, ElementType attackType) {
        // Check if the entity has properties.
        if (this.getFightProperties() == null || !hasFightProperty(FightProperty.FIGHT_PROP_CUR_HP)) {
            return;
        }

        // Invoke entity damage event.
        EntityDamageEvent event =
                new EntityDamageEvent(this, amount, attackType, this.getScene().getEntityById(killerId));
        event.call();
        if (event.isCanceled()) {
            return; // If the event is canceled, do not damage the entity.
        }

        float curHp = getFightProperty(FightProperty.FIGHT_PROP_CUR_HP);
        if (curHp != Float.POSITIVE_INFINITY && !lockHP || lockHP && curHp <= event.getDamage()) {
            // Add negative HP to the current HP property.
            this.addFightProperty(FightProperty.FIGHT_PROP_CUR_HP, -(event.getDamage()));
        }

        this.lastAttackType = attackType;

        // Check if dead
        if (this.getFightProperty(FightProperty.FIGHT_PROP_CUR_HP) <= 0f) {
            this.setFightProperty(FightProperty.FIGHT_PROP_CUR_HP, 0f);
            this.isDead = true;
        }

        this.runLuaCallbacks(event);

        // Packets
        this.getScene()
                .broadcastPacket(
                        new PacketEntityFightPropUpdateNotify(this, FightProperty.FIGHT_PROP_CUR_HP));

        // Check if dead.
        if (this.isDead) {
            this.getScene().killEntity(this, killerId);
        }
    }

    /**
     * Runs the Lua callbacks for {@link EntityDamageEvent}.
     *
     * @param event The damage event.
     */
    public void runLuaCallbacks(EntityDamageEvent event) {
        if (entityController != null) {
            entityController.onBeHurt(this, event.getAttackElementType(), true); // todo is host handling
        }
    }

    /**
     * Move this entity to a new position.
     *
     * @param position The new position.
     * @param rotation The new rotation.
     */
    public void move(Position position, Position rotation) {
        // Set the position and rotation.
        this.getPosition().set(position);
        this.getRotation().set(rotation);
    }

    /**
     * Called when a player interacts with this entity
     *
     * @param player Player that is interacting with this entity
     * @param interactReq Interact request protobuf data
     */
    public void onInteract(Player player, GadgetInteractReq interactReq) {}

    /** Called when this entity is added to the world */
    public void onCreate() {}

    public void onRemoved() {}

    private int[] parseCountRange(String range) {
        var split = range.split(";");
        if (split.length == 1)
            return new int[] {Integer.parseInt(split[0]), Integer.parseInt(split[0])};
        return new int[] {Integer.parseInt(split[0]), Integer.parseInt(split[1])};
    }

    public boolean dropSubfieldItem(int dropId) {
        var drop = GameData.getDropSubfieldMappingMap().get(dropId);
        if (drop == null) return false;
        var dropTableEntry = GameData.getDropTableExcelConfigDataMap().get(drop.getItemId());
        if (dropTableEntry == null) return false;

        Int2ObjectMap<Integer> itemsToDrop = new Int2ObjectOpenHashMap<>();
        switch (dropTableEntry.getRandomType()) {
            case 0: // select one
                {
                    int weightCount = 0;
                    for (var entry : dropTableEntry.getDropVec()) weightCount += entry.getWeight();

                    int randomValue = new Random().nextInt(weightCount);

                    weightCount = 0;
                    for (var entry : dropTableEntry.getDropVec()) {
                        if (randomValue >= weightCount && randomValue < (weightCount + entry.getWeight())) {
                            var countRange = parseCountRange(entry.getCountRange());
                            itemsToDrop.put(
                                    entry.getItemId(),
                                    Integer.valueOf((new Random().nextBoolean() ? countRange[0] : countRange[1])));
                        }
                    }
                }
                break;
            case 1: // Select various
                {
                    for (var entry : dropTableEntry.getDropVec()) {
                        if (entry.getWeight() < new Random().nextInt(10000)) {
                            var countRange = parseCountRange(entry.getCountRange());
                            itemsToDrop.put(
                                    entry.getItemId(),
                                    Integer.valueOf((new Random().nextBoolean() ? countRange[0] : countRange[1])));
                        }
                    }
                }
                break;
        }

        for (var entry : itemsToDrop.int2ObjectEntrySet()) {
            var item =
                    new EntityItem(
                            scene,
                            null,
                            GameData.getItemDataMap().get(entry.getIntKey()),
                            getPosition().nearby2d(1f).addY(0.5f),
                            entry.getValue(),
                            true);

            scene.addEntity(item);
        }

        return true;
    }

    public boolean dropSubfield(String subfieldName) {
        var subfieldMapping = GameData.getSubfieldMappingMap().get(getEntityTypeId());
        if (subfieldMapping == null || subfieldMapping.getSubfields() == null) return false;

        for (var entry : subfieldMapping.getSubfields()) {
            if (entry.getSubfieldName().compareTo(subfieldName) == 0) {
                return dropSubfieldItem(entry.getDrop_id());
            }
        }

        return false;
    }

    public void onTick(int sceneTime) {
        if (entityController != null) {
            entityController.onTimer(this, sceneTime);
        }
    }

    public int onClientExecuteRequest(int param1, int param2, int param3) {
        if (entityController != null) {
            return entityController.onClientExecuteRequest(this, param1, param2, param3);
        }
        return 0;
    }

    /**
     * Called when this entity dies
     *
     * @param killerId Entity id of the entity that killed this entity
     */
    public void onDeath(int killerId) {
        // Invoke entity death event.
        EntityDeathEvent event = new EntityDeathEvent(this, killerId);
        event.call();

        // Run Lua callbacks.
        if (entityController != null) {
            entityController.onDie(this, getLastAttackType());
        }
    }

    /** Invoked when a global ability value is updated. */
    public void onAbilityValueUpdate() {
        // Does nothing.
    }

    public abstract SceneEntityInfo toProto();

    @Override
    public String toString() {
        return "Entity ID: %s; Group ID: %s; Config ID: %s"
                .formatted(this.getId(), this.getGroupId(), this.getConfigId());
    }
}

File: src/main/java/emu/grasscutter/game/entity/GameEntity.kt
package emu.grasscutter.game.entity

import emu.grasscutter.game.props.FightProperty
import emu.grasscutter.game.props.LifeState
import emu.grasscutter.game.props.PlayerProperty
import emu.grasscutter.game.world.Scene
import emu.grasscutter.game.world.World
import emu.grasscutter.net.proto.AbilitySyncStateInfoOuterClass.AbilitySyncStateInfo
import emu.grasscutter.net.proto.AnimatorParameterValueInfoOuterClass.AnimatorParameterValueInfo
import emu.grasscutter.net.proto.AnimatorParameterValueInfoPairOuterClass.AnimatorParameterValueInfoPair
import emu.grasscutter.net.proto.EntityAuthorityInfoOuterClass.EntityAuthorityInfo
import emu.grasscutter.net.proto.EntityClientDataOuterClass.EntityClientData
import emu.grasscutter.net.proto.EntityEnvironmentInfoOuterClass.EntityEnvironmentInfo
import emu.grasscutter.net.proto.EntityRendererChangedInfoOuterClass.EntityRendererChangedInfo
import emu.grasscutter.net.proto.FightPropPairOuterClass.FightPropPair
import emu.grasscutter.net.proto.MotionInfoOuterClass.MotionInfo
import emu.grasscutter.net.proto.MotionStateOuterClass.MotionState
import emu.grasscutter.net.proto.PropPairOuterClass.PropPair
import emu.grasscutter.net.proto.ProtEntityTypeOuterClass.ProtEntityType
import emu.grasscutter.net.proto.SceneEntityAiInfoOuterClass.SceneEntityAiInfo
import emu.grasscutter.net.proto.SceneEntityInfoOuterClass.SceneEntityInfo
import emu.grasscutter.net.proto.ServerBuffOuterClass.ServerBuff
import emu.grasscutter.net.proto.VectorOuterClass
import emu.grasscutter.utils.Position
import emu.grasscutter.utils.ProtoHelper

abstract class GameEntity {

	abstract val scene: Scene
	abstract val position: Position
	abstract val rotation: Position
	abstract val entityType: ProtEntityType

	val motionInfo: MotionInfo.Builder by lazy {
		MotionInfo.newBuilder()
			.setPos(position.toProto())
			.setRot(rotation.toProto())
			.setSpeed(VectorOuterClass.Vector.newBuilder())
			.setState(motionState)
	}


	var id = -1
	val name = SettingContainer("none")
	val propPairList = SettingContainer(arrayListOf(PropPair.newBuilder()
		.setType(PlayerProperty.PROP_LEVEL.id)
		.setPropValue(ProtoHelper.newPropValue(PlayerProperty.PROP_LEVEL, 1)).build()), changed=true)
	open val fightPropPairList = SettingContainer(mutableMapOf<Int, Float>(), changed=true)
	val lifeState = SettingContainer(LifeState.LIFE_ALIVE)
	val animatorParaList = SettingContainer(mutableMapOf<Int,AnimatorParameterValueInfo>(), changed=true)
	val lastMoveSceneTimeMs = SettingContainer(-1)
	val lastMoveReliableSeq = SettingContainer(-1)
	val entityClientData = SettingContainer(EntityClientData.newBuilder(), changed=true)
	val entityEnvironmentInfoList = SettingContainer(arrayListOf<EntityEnvironmentInfo>())
	val entityAuthorityInfo = SettingContainer(EntityAuthorityInfo.newBuilder()
		.setAbilityInfo(AbilitySyncStateInfo.newBuilder())
		.setRendererChangedInfo(EntityRendererChangedInfo.newBuilder())
		.setAiInfo(SceneEntityAiInfo.newBuilder().setIsAiOpen(true).setBornPos(VectorOuterClass.Vector.newBuilder()))
		.setBornPos(VectorOuterClass.Vector.newBuilder()), changed=true)
	val tagList = SettingContainer(arrayListOf<String>())
	val serverBuffList = SettingContainer(arrayListOf<ServerBuff>())

	var motionState: MotionState = MotionState.MOTION_NONE
		set(state) {
			this.motionInfo.state = state
			field = state
		}


	open fun getFightProperty(propType: FightProperty): Float {
		return fightPropPairList.value[propType.id]?:0f
	}

	open fun setFightProperty(propType: FightProperty, propValue: Float) {
		fightPropPairList.value[propType.id] = propValue
	}

	open fun addFightProperty(propType: FightProperty, propValue: Float) {
		fightPropPairList.value[propType.id] = fightPropPairList.value.getOrDefault(propType.id, 0f) + propValue
	}

	fun setLevel(level: Int) {
		propPairList.value.removeIf { it.type == PlayerProperty.PROP_LEVEL.id }
		val pair = PropPair.newBuilder()
			.setType(PlayerProperty.PROP_LEVEL.id)
			.setPropValue(ProtoHelper.newPropValue(PlayerProperty.PROP_LEVEL, level))
			.build()
		propPairList.value.add(pair)
	}


	fun getWorld(): World {
		return scene.world
	}

	abstract fun toProto(): SceneEntityInfo


	open fun getInitProto(): SceneEntityInfo.Builder {

		val authority = this.entityAuthorityInfo

		val entityInfo = SceneEntityInfo.newBuilder()
			.setEntityType(this.entityType)
			.setEntityId(id)
			.setMotionInfo(this.motionInfo)
			.setEntityClientData(this.entityClientData.value)
			.setEntityAuthorityInfo(authority.value)
			.addAllPropList(this.propPairList.value)


		animatorParaList.ifChangedAlsoElse(
			{ entityInfo.addAllAnimatorParaList(this.animatorParaList.value.map { (k, v) -> AnimatorParameterValueInfoPair.newBuilder().setNameId(k).setAnimatorPara(v).build()}) },
			{ entityInfo.addAllAnimatorParaList(arrayListOf(AnimatorParameterValueInfoPair.newBuilder().build()))})

		fightPropPairList.ifChangedAlsoElse(
			{ entityInfo.addAllFightPropList(it.map { (k, v) -> FightPropPair.newBuilder().setPropType(k).setPropValue(v).build() }) },
			{ entityInfo.addAllFightPropList(arrayListOf(FightPropPair.newBuilder().build())) })


		lifeState.ifChangedAlso { entityInfo.lifeState = it.value }
		lastMoveReliableSeq.ifChangedAlso(entityInfo::setLastMoveReliableSeq)
		lastMoveSceneTimeMs.ifChangedAlso(entityInfo::setLastMoveSceneTimeMs)
		entityEnvironmentInfoList.ifChangedAlso(entityInfo::addAllEntityEnvironmentInfoList)
		tagList.ifChangedAlso(entityInfo::addAllTagList)
		serverBuffList.ifChangedAlso(entityInfo::addAllServerBuffList)


		return entityInfo
	}

	open fun onDeath(killerId: Int) {

	}


}

abstract class EntityLife(level: Int): GameEntity() {

	init {
		this.lifeState.set(LifeState.LIFE_ALIVE)
		setLevel(level)
	}

	open fun isAlive(): Boolean {
		return this.getFightProperty(FightProperty.FIGHT_PROP_CUR_HP) > 0f
	}

	override fun getInitProto(): SceneEntityInfo.Builder {
		this.lifeState.set(if (isAlive()) LifeState.LIFE_ALIVE else LifeState.LIFE_DEAD)

		return super.getInitProto()
	}

}

class SettingContainer<T>(val defaultValue: T, currentValue: T? = null, changed: Boolean = false) {

	var value: T = currentValue ?: defaultValue
		private set

	var hasChanged: Boolean = (currentValue != null || changed)
	

	fun set(change: T?) {
		change?.also {
			value = it
			hasChanged = true
		}
	}

	fun hasChanged(): Boolean = hasChanged

	fun reset() {
		this.value = defaultValue
	}

	inline fun ifChangedAlso(func: (T) -> Unit) {
		if(hasChanged) func(value)
	}

	inline fun ifChangedAlsoElse(func: (T) -> Unit, ifElse: () -> Unit) {
		if(hasChanged) {
			func(value)
		} else {
			ifElse()
		}
	}

	inline fun <R> ifChangedLet(func: (T) -> R) {
		if (hasChanged) func(value)
	}

}
